File: backend/src/test/java/com/kiwi/metrics/MetricsFactory.java
package com.kiwi.metrics;

import com.kiwi.features.metrics.data.MetricsDTO;

import java.time.LocalDate;
import java.util.HashSet;
import java.util.Random;
import java.util.Set;

import static com.kiwi.common.utils.FormatUtils.formatDate;

public class MetricsFactory {
    private static final Random RANDOM = new Random();

    public static MetricsDTO generateRandomValidMetricDTO() {
        return new MetricsDTO(
            getRandomDate(),
            0,
            getRandomTimeSeconds(1, 2),
            0,
            getRandomTimeSeconds(3, 4)
        );
    }

    public static MetricsDTO generateRandomInvalidMetricDTO() {
        return new MetricsDTO(
            getRandomDate(),
            0,
            -getRandomTimeSeconds(1, 2),
            0,
            -getRandomTimeSeconds(3, 4)
        );
    }
    
    public static Set<MetricsDTO> generateRandomMetricsSet(int size, boolean getValidValues) {
        Set<MetricsDTO> metricsDTOSet = new HashSet<>();
        for (int i = 0; i < size; i++) {
            metricsDTOSet.add(getValidValues ? generateRandomValidMetricDTO() : generateRandomInvalidMetricDTO());
        }
        return metricsDTOSet;
    }
    
    private static String getRandomDate() {
        int startYear = 2025;
        int endYear = 2026;

        int year = RANDOM.nextInt(endYear - startYear + 1) + startYear;
        int month = RANDOM.nextInt(12) + 1;
        int dayOfMonth = RANDOM.nextInt(28) + 1;

        return formatDate(LocalDate.of(year, month, dayOfMonth));
    }
    
    private static Integer getRandomTimeSeconds(Integer fromHours, Integer untilHours) {
        return RANDOM.nextInt((untilHours - fromHours) * 60 * 60) + fromHours;
    }
}

File: mobile/app/src/main/java/com/bellako/kiwi/features/metrics/model/MetricsFactory.kt
package com.bellako.kiwi.features.metrics.model

import android.os.Build
import androidx.annotation.RequiresApi
import com.bellako.kiwi.common.utils.SECONDS_IN_HOUR
import com.bellako.kiwi.features.metrics.data.MetricsDTO
import java.time.LocalDate
import java.time.format.DateTimeFormatter
import kotlin.random.Random

@Suppress("MagicNumber")
@RequiresApi(Build.VERSION_CODES.O)
object MetricsFactory {
    private val formatter = DateTimeFormatter.ISO_LOCAL_DATE

    fun generateRandomValidMetricDTO(): MetricsDTO =
        MetricsDTO(
            date = getRandomDate(),
            maxGoodTimeSeconds = getRandomTimeSeconds(5, 6),
            currentGoodTimeSeconds = getRandomTimeSeconds(1, 2),
            maxBadTimeSeconds = getRandomTimeSeconds(5, 6),
            currentBadTimeSeconds = getRandomTimeSeconds(3, 4),
        )

    fun generateRandomInvalidMetricDTO(): MetricsDTO =
        MetricsDTO(
            date = getRandomDate(),
            maxGoodTimeSeconds = -getRandomTimeSeconds(5, 6),
            currentGoodTimeSeconds = -getRandomTimeSeconds(1, 2),
            maxBadTimeSeconds = -getRandomTimeSeconds(5, 6),
            currentBadTimeSeconds = -getRandomTimeSeconds(3, 4),
        )

    fun generateRandomMetricsSet(
        size: Int,
        valid: Boolean,
    ): Set<MetricsDTO> =
        (1..size)
            .map {
                if (valid) generateRandomValidMetricDTO() else generateRandomInvalidMetricDTO()
            }.toSet()

    private fun getRandomDate(): String {
        val year = Random.nextInt(2025, 2027)
        val month = Random.nextInt(1, 13)
        val day = Random.nextInt(1, 29)

        return LocalDate.of(year, month, day).format(formatter)
    }

    private fun getRandomTimeSeconds(
        fromHours: Int,
        untilHours: Int,
    ): Int = Random.nextInt(fromHours * SECONDS_IN_HOUR, untilHours * SECONDS_IN_HOUR)
}
